crates/sui-axelar-cgp/move/sources/channel.move

```diff
@@ -0,0 +1,131 @@
+    const EDuplicateMessage: u64 = 2;
+
+    struct Channel<T: store> has store {
+        /// Unique ID of the target object which allows message targeting
+        /// by comparing against `id_bytes`.
+        id: UID,
+        /// Messages processed by this object. To make system less
+        /// centralized, and spread the storage + io costs across multiple
+        /// destinations, we can track every `Channel`'s messages.
+        messages: VecSet<vector<u8>>,
+        /// Additional field to optionally use as metadata for the Channel
+        /// object improving identification and uniqueness of data.
+        /// Can store any struct that has `store` ability (including other
+        /// objects - eg Capabilities).
+        data: T
+    }
+
+    /// Emitted when a new message is sent from the SUI network.
+    struct MessageSent has copy, drop {
+        source: vector<u8>,
+        destination: vector<u8>,
+        destination_address: vector<u8>,
+        payload: vector<u8>,
+    }
+
+    /// Create new `Channel<T>` object. Anyone can create their own `Channel` to target
+    /// from the outside and there's no limitation to the data stored inside it.
+    ///
+    /// `copy` ability is required to disallow asset locking inside the `Channel`.
+    public fun create_channel<T: store>(t: T, ctx: &mut TxContext): Channel<T> {
+        Channel {
+            id: object::new(ctx),
+            messages: vec_set::empty(),
+            data: t
+        }
+    }
+
+    /// Destroy a `Channel<T>` releasing the T. Not constrained and can be performed
+    /// by any party as long as they own a Channel.
+    public fun destroy_channel<T: store>(self: Channel<T>): T {
+        let Channel { id, messages: _, data } = self;
+        object::delete(id);
+        data
+    }
+
+    /// Send a message to another chain. Supply the event data and the
+    /// destination chain.
+    ///
+    /// Event data is collected from the Channel (eg ID of the source and
+    /// source_chain is a constant).
+    public fun send_message<T: store>(
+        t: &mut Channel<T>,
+        destination: vector<u8>,
+        destination_address: vector<u8>,
+        payload: vector<u8>
+    ) {
+        sui::event::emit(MessageSent {
+            source: bcs::to_bytes(&t.id),
+            destination,
+            destination_address,
+            payload,
+        })
+    }
+
+    /// By using &mut here we make sure that the object is not in the freeze
+    /// state and the owner has full access to the target.
+    ///
+    /// Most common scenario would be to target a shared object, however this
+    /// messaging system allows sending private messages which can be consumed
+    /// by single-owner targets.
+    ///
+    /// For Capability-locking, a mutable reference to the `Channel.data` field is
+    /// returned; the rest are the fields of the `Message`.
+    public fun consume_message<T: store>(
```

Comment:
Can we make Message a hot potato and weaken the link between Axelar and consume_message? Similar to TransferPolicy + TransferRequest.